storage/innobase/fil/fil0crypt.cc

```diff
@@ -1117,11 +1122,22 @@ static bool fil_crypt_space_needs_rotation(rotate_thread_t *state,
         crypt_data->encryption, crypt_data->min_key_version,
         key_state->key_version, key_state->rotate_key_age);
 
-    if (need_key_rotation && crypt_data->rotate_state.active_threads > 0 &&
-        crypt_data->rotate_state.next_offset >
-            crypt_data->rotate_state.max_offset) {
-      break;  // the space is already being processed and there are no more
-              // pages to rotate
+    if (need_key_rotation) {
+      if (crypt_data->rotate_state.active_threads == 0) {
+        if (key_state->key_version == ENCRYPTION_KEY_VERSION_INVALID) {
+          ut_ad(is_unenc_to_enc_rotation(*crypt_data));
+          // we are asked to encrypt table with encryption_key_id assigned, but
```

Comment:
but active_threads are zero, which means encryption threads are zero?

What is the point in creating this tablespace key? Who is going to encrypt the table and use it?

Also isn't it better to do this key creation at place where tablespace encryption is enabled?

Reply:
Question: Wjy do we do allow encryption_key_id for unencrypted tablespaces?
Answer from slack " Without this if encryption threads are enabled they would use only the default encryption key to encrypt tables. So no way of saying that a particular tablespace should be encrypted with different encryption key by encryption threads. With that you can overwrite this behavior - encryption threads will use the assigned encryption_key_id."

Reply:
>     1. active_threads are 0 which means that there is no encryption threads currently processing the tablespace. So we are the first thread checking if this tablespace needs rotation. Encryption threads are (of course) on for this.
> 
>     2. As in 1) - first encryption thread has figured out that this tablespace needs rotation. Since it is unencrypted => encrypted rotation we need to make sure that key id that is to be used during rotation exists and create it if it doesn't.
> 
>     3. Normally it is created when encryption key id is specified. This is to mitigate the problem when keyring file was changed and encryption key id does not exist. For instance encryption threads are disabled, nothing is encrypted, but the user create tablespaces with encryption_key_id. Then he deletes the keyring. After restarts all is working well. Then he loads a new keyring and decides to enable encryption threads. Stuff like that.

Thanks for answering the questions :+1: